pkgs/development/compilers/llvm/common/clang-tools/wrapper

```diff
@@ -2,26 +2,43 @@
 
 buildcpath() {
   local path after
+
   while (( $# )); do
     case $1 in
         -isystem)
             shift
             path=$path${path:+':'}$1
             ;;
+
         -idirafter)
             shift
             after=$after${after:+':'}$1
             ;;
     esac
+
     shift
   done
+
   echo $path${after:+':'}$after
 }
 
-export CPATH=${CPATH}${CPATH:+':'}$(buildcpath ${NIX_CFLAGS_COMPILE} \
-                                               $(<@clang@/nix-support/libc-cflags)):@clang@/resource-root/include
-export CPLUS_INCLUDE_PATH=${CPLUS_INCLUDE_PATH}${CPLUS_INCLUDE_PATH:+':'}$(buildcpath ${NIX_CFLAGS_COMPILE} \
-                                                                                      $(<@clang@/nix-support/libcxx-cxxflags) \
-                                                                                      $(<@clang@/nix-support/libc-cflags)):@clang@/resource-root/include
+# When user passes `--query-driver`, avoid extending `CPATH` et al, since we
+# don't want to "infect" user-specified toolchain and headers with our stuff.
+extendcpath=true
+
+for arg in "$@"; do
+  if [[ "${arg}" == \-\-query\-driver* ]]; then
+    extendcpath=false
+  fi
+done
+
+if [ "$extendcpath" = true ]; then
```

Comment:
[nit] `[ -n "$extendcpath" ]`, with possible values for extendcpath being either the empty string or 1, would be more idiomatic for shell.